nmigen_boards/orangecrab_r0_2.py

```diff
@@ -0,0 +1,109 @@
+                           attrs=Attrs(IO_STANDARD="LVCMOS33")
+        ),
+
+        Resource("ddr3", 0,
+            Subsignal("rst",     PinsN("L18", dir="o")),
+            Subsignal("clk",     DiffPairs("J18", "K18", dir="o"), Attrs(IO_TYPE="SSTL135D_I")),
+            Subsignal("clk_en",  Pins("D18", dir="o")),
+            Subsignal("cs",      PinsN("A12", dir="o")),
+            Subsignal("we",      PinsN("B12", dir="o")),
+            Subsignal("ras",     PinsN("C12", dir="o")),
+            Subsignal("cas",     PinsN("D13", dir="o")),
+            Subsignal("a",       Pins("C4 D2 D3 A3 A4 D4 C3 B2 B1 D1 A7 C2 B6 C1 A2 C7", dir="o")),
+            Subsignal("ba",      Pins("P5 N3 M3", dir="o")),
+            Subsignal("dqs",     DiffPairs("G18 H17", "B15 A16", dir="io"),
+                      Attrs(IO_TYPE="SSTL135D_I", TERMINATION="OFF",
+                      DIFFRESISTOR="100")),
+            Subsignal("dq",      Pins("C17 D15 B17 C16 A15 B13 A17 A13 F17 F16 G15 F15 J16 C18"
+                                      "H16 F18", dir="io"), Attrs(TERMINATION="75")),
+            Subsignal("dm",      Pins("G16 D16", dir="o")),
+            Subsignal("odt",     Pins("C13", dir="o")),
+            # top up pins, leave these at their default value
+            Subsignal("vcc_virtual", PinsN("K16 D17 K15 K17 B18 C6", dir="o"),
+                      Attrs(IO_TYPE="SSTL135_II")),
+            Subsignal("gnd_virtual", Pins("L15 L16", dir="o"), Attrs(IO_TYPE="SSTL135_II")),
+            Attrs(IO_TYPE="SSTL135_I", SLEWRATE="FAST")
+        ),
+
+        Resource("adc", 0,
+            Subsignal("ctrl",     Pins("G1 F1", dir="o")),
+            Subsignal("mux",      Pins("F4 F3 F2 H1", dir="o")),
+            Subsignal("sense_hi", Pins("H3", dir="i")),
+            Subsignal("sense_lo", Pins("G3", dir="i")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        *SDCardResources(0, dat0="J1", dat1="K3", dat2="L3", dat3="M1", clk="K1", cmd="K2", cd="L1",
+            attrs=Attrs(IO_TYPE="LVCMOS33", SLEWRATE="FAST")
+        ),
+
+        Resource("usb", 0,
+            Subsignal("d_p",    Pins("N1", dir="io")),
+            Subsignal("d_m",    Pins("M2", dir="io")),
+            Subsignal("pullup", Pins("N2", dir="o")),
+            Attrs(IO_TYPE="LVCMOS33")
+        )
+    ]
+    connectors = [
+        Connector("io", 0, "N17 M18 - - - B10 B9 - - C8 B8 A8 H2 J2"),
+        Connector("a", 0, "L4 N3 N4 H4 G4"),
+        Connector("spi", 0, {
+            "miso": "N15",
+            "mosi": "N16",
+            "sck":  "R17"
+        }),
+        Connector("i2c", 0, {
+            "scl": "C9",
+            "sda": "C10"
+        })
+    ]
+
+    def toolchain_prepare(self, fragment, name, **kwargs):
+        overrides = dict(ecppack_opts="--compress --freq 38.8")
+        overrides.update(kwargs)
+
+        return super().toolchain_prepare(fragment, name, **overrides, **kwargs)
```

Comment:
You're passing both `overrides` and `kwargs` here, which will break if `kwargs` includes `ecppack_opts`:

```python
>>> def foo(**kw): pass
...
>>> x = dict(a=1)
>>> y = dict(a=2)
>>> foo(**x,**y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() got multiple values for keyword argument 'a'
```

Reply:
Ah, I only tried with NMIGEN_ECPPACK_OPTS which works differently than I thought.